tests/unit/test_cli.py

```diff
@@ -189,6 +191,40 @@ def side_effect():
         mock_daemon_start.assert_called_once()
 
 
+    def test_ensure_directory_exists_not_exists(self):
+        with tempfile.TemporaryDirectory() as tempdir:
+            test_dir = pathlib.Path(tempdir).joinpath("test_ensure")
+            cli.ensure_directory_exists(str(test_dir))
+            self.assertTrue(test_dir.is_dir())
+
+
+    def test_ensure_directory_exists_not_writable(self):
+        if os.getuid() == 0:
+            logger = logging.getLogger(__name__)
+            logger.warning("test_ensure_directory_exists_not_writable cannot fail when run as root")
+            # prefer un-failable test to un-passable test
+            return
+
+        with tempfile.TemporaryDirectory() as tempdir:
+            test_dir = pathlib.Path(tempdir).joinpath("test_ensure")
+            test_dir.mkdir(mode=0o555)
+            self.assertRaises(ConfigWriteError, cli.ensure_directory_exists, str(test_dir))
+
+
+    def test_ensure_directory_exists_success(self):
+        with tempfile.TemporaryDirectory() as tempdir:
+            test_dir = pathlib.Path(tempdir).joinpath("test_ensure")
+            test_dir.mkdir(mode=0o777)
+            cli.ensure_directory_exists(str(test_dir))
```

Comment:
I could put the call to `ensure_directory_exists` in a try/except catching just the `NotADirectoryError` and `ConfigWriteError` and explicitly failing on those.

Reply:
hmm... what is the goal of this test?

Reply:
Issue #3270 asks for 4 test cases. One of them is 

> dir exists and is writable

Reply:
maybe your test can try creating a file in the expected directory and catch the error since it doesn't exist, then call `ensure_directory_exists` and try writing the file again this time without catching the error (since it should work), this would clearly show what didn't work before calling the function and what now works after calling the function

Reply:
`test_ensure_directory_exists_not_exists` does basically that already. I wrote it to cover the requested test case

> download directory does not exist

`test_ensure_directory_exists_success` was meant to test the case where the directory already exists and is writable, and so `ensure_directory_exists` would actually do nothing (but also not fail).

If you don't think that case is useful, I could just remove `test_ensure_directory_exists_success` altogether, and `test_ensure_directory_exists_not_exists` would still be there to cover the case you're talking about.